javascript/ql/src/semmle/javascript/security/dataflow/TaintedPath.qll

```diff
@@ -29,15 +134,373 @@ module TaintedPath {
+      DataFlow::FlowLabel dstlabel
+    ) {
+      // path.normalize() and similar
+      exists(NormalizingPathCall call |
+        src = call.getInput() and
+        dst = call.getOutput() and
+        dstlabel = Label::toPosixPath(srclabel).toNormalized()
+      )
+      or
+      // path.resolve() and similar
+      exists(ResolvingPathCall call |
+        src = call.getInput() and
+        dst = call.getOutput() and
+        srclabel = anyLabel() and
+        dstlabel.(Label::PosixPath).isAbsolute() and
+        dstlabel.(Label::PosixPath).isNormalized()
+      )
+      or
+      // path.relative() and similar
+      exists(NormalizingRelativePathCall call |
+        src = call.getInput() and
+        dst = call.getOutput() and
+        dstlabel.(Label::PosixPath).isRelative() and
+        dstlabel.(Label::PosixPath).isNormalized()
+      )
+      or
+      // path.dirname() and similar
+      exists(PreservingPathCall call |
+        src = call.getInput() and
+        dst = call.getOutput() and
+        preserveLabel(srclabel, dstlabel)
+      )
+      or
+      // path.join()
+      exists(DataFlow::CallNode join, int n |
+        join = DataFlow::moduleMember("path", "join").getACall()
+      |
+        src = join.getArgument(n) and
+        dst = join and
+        (
+          // If the initial argument is tainted, just normalize it. It can be relative or absolute.
+          n = 0 and
+          dstlabel = Label::toPosixPath(srclabel).toNormalized()
+          or
+          // For later arguments, the flow label depends on whether the first argument is absolute or relative.
+          // If in doubt, we assume it is absolute.
+          n > 0 and
+          Label::toPosixPath(srclabel).canContainDotDotSlash() and
+          dstlabel.(Label::PosixPath).isNormalized() and
+          if isRelative(join.getArgument(0).getStringValue())
+          then dstlabel.(Label::PosixPath).isRelative()
+          else dstlabel.(Label::PosixPath).isAbsolute()
+        )
+      )
+      or
+      // String concatenation - behaves like path.join() except without normalization
+      exists(DataFlow::Node operator, int n |
+        StringConcatenation::taintStep(src, dst, operator, n)
+      |
+        // use ordinary taint flow for the first operand
+        n = 0 and
+        preserveLabel(srclabel, dstlabel)
+        or
+        n > 0 and
+        Label::toPosixPath(srclabel).canContainDotDotSlash() and
+        dstlabel.(Label::PosixPath).isNonNormalized() and // The ../ is no longer at the beginning of the string.
+        (
+          if isRelative(StringConcatenation::getOperand(operator, 0).getStringValue())
+          then dstlabel.(Label::PosixPath).isRelative()
+          else dstlabel.(Label::PosixPath).isAbsolute()
+        )
+      )
+    }
+  }
+
+  /**
+   * Holds if `s` is a relative path.
+   */
+  bindingset[s]
+  private predicate isRelative(string s) { not s = "/" + any(string q) }
```

Comment:
`not s.charAt(0) = "/"`?

Reply:
👍 Didn't know https://jira.semmle.com/browse/QL-45 had been fixed.